File: hash/go/imagehash/internal/bitpack/bitpack_test.go
package bitpack_test

import (
	"testing"

	"github.com/wmetcalf/rosetta-image-hash/go/imagehash/internal/bitpack"
)

func TestPack4x4Pattern(t *testing.T) {
	bits := [][]bool{
		{true, false, true, false},
		{false, true, false, true},
		{true, true, true, true},
		{false, false, false, false},
	}
	if got := bitpack.Pack(bits); got != "a5f0" {
		t.Errorf("got %q, want a5f0", got)
	}
}

func TestPackAllOnes8x8(t *testing.T) {
	bits := make([][]bool, 8)
	for y := range bits {
		bits[y] = make([]bool, 8)
		for x := range bits[y] {
			bits[y][x] = true
		}
	}
	if got := bitpack.Pack(bits); got != "ffffffffffffffff" {
		t.Errorf("got %q", got)
	}
}

func TestPackAllZeros8x8(t *testing.T) {
	bits := make([][]bool, 8)
	for y := range bits {
		bits[y] = make([]bool, 8)
	}
	if got := bitpack.Pack(bits); got != "0000000000000000" {
		t.Errorf("got %q", got)
	}
}

func TestUnpackSquareIsInverse(t *testing.T) {
	expected := [][]bool{
		{true, false, true, false},
		{false, true, false, true},
		{true, true, true, true},
		{false, false, false, false},
	}
	got, err := bitpack.UnpackSquare("a5f0")
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if len(got) != 4 {
		t.Fatalf("rows: %d", len(got))
	}
	for y := range expected {
		for x := range expected[y] {
			if got[y][x] != expected[y][x] {
				t.Errorf("(%d,%d): got %v, want %v", y, x, got[y][x], expected[y][x])
			}
		}
	}
}

func TestUnpackFlat(t *testing.T) {
	got, err := bitpack.UnpackFlat("00000000000", 3)
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if len(got) != 14 || len(got[0]) != 3 {
		t.Fatalf("shape: %dx%d", len(got), len(got[0]))
	}
	for y := range got {
		for x := range got[y] {
			if got[y][x] {
				t.Errorf("expected all false at (%d,%d)", y, x)
			}
		}
	}
}

func TestUnpackSquareNonSquareLengthErrors(t *testing.T) {
	if _, err := bitpack.UnpackSquare("12345"); err == nil {
		t.Errorf("expected error for non-square hex")
	}
}

func TestUnpackInvalidCharsErrors(t *testing.T) {
	if _, err := bitpack.UnpackSquare("xyz!"); err == nil {
		t.Errorf("expected error for invalid chars")
	}
}

File: hash/java/src/test/java/io/github/wmetcalf/rosettasquint/hash/group1/BitPackTest.java
package io.github.wmetcalf.rosettasquint.hash.group1;

import io.github.wmetcalf.rosettasquint.hash.internal.BitPack;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class BitPackTest {
    @Test
    void pack4x4Pattern() {
        boolean[][] bits = {
                {true, false, true, false},
                {false, true, false, true},
                {true, true, true, true},
                {false, false, false, false},
        };
        assertEquals("a5f0", BitPack.pack(bits));
    }

    @Test
    void packAllOnes8x8() {
        boolean[][] bits = new boolean[8][8];
        for (int y = 0; y < 8; y++) for (int x = 0; x < 8; x++) bits[y][x] = true;
        assertEquals("ffffffffffffffff", BitPack.pack(bits));
    }

    @Test
    void packAllZeros8x8() {
        boolean[][] bits = new boolean[8][8];
        assertEquals("0000000000000000", BitPack.pack(bits));
    }

    @Test
    void unpackSquareIsInverse() {
        boolean[][] bits = {
                {true, false, true, false},
                {false, true, false, true},
                {true, true, true, true},
                {false, false, false, false},
        };
        boolean[][] roundTrip = BitPack.unpackSquare("a5f0");
        assertEquals(4, roundTrip.length);
        for (int y = 0; y < 4; y++) {
            assertArrayEquals(bits[y], roundTrip[y]);
        }
    }

    @Test
    void unpackFlat() {
        boolean[][] bits = BitPack.unpackFlat("00000000000", 3);
        assertEquals(14, bits.length);
        assertEquals(3, bits[0].length);
        for (int i = 0; i < 14; i++)
            for (int j = 0; j < 3; j++)
                assertEquals(false, bits[i][j]);
    }

    @Test
    void unpackSquareNonSquareLengthThrows() {
        assertThrows(IllegalArgumentException.class, () -> BitPack.unpackSquare("12345"));
    }
}
